### PR TITLE
[azure-core/corehttp] Add JSONL and SSE streaming support

### PR DESCRIPTION
**Pull request by l0lawrence**
Adds streaming support to **both azure-core (`azure.core.streaming`) and corehttp (`corehttp.streaming`)**, kept in sync so the two implementations stay identical (only the `rest`-path docstrings differ).

### What's included (per package)

**Core**
- `Stream` / `AsyncStream` — decoder-agnostic iterators (`__iter__`/`__aiter__` return `self`)
- `deserialization_callback` receives `(response, event)` so generated code can support the `cls` custom-deserializer pattern

**JSONL**
- `JSONLDecoder` / `AsyncJSONLDecoder` with incremental UTF-8 line decoding
- Lines are split only on `\n` (with `\r\n` tolerated), **not** `str.splitlines()`, so a JSON string value containing other Unicode line boundaries (e.g. `\u2028`, `\x85`) is not split mid-record

**SSE (Server-Sent Events)**
- `ServerSentEvent` value object (`event`, `data`, `id`, `retry`) plus `SSEDecoder` / `AsyncSSEDecoder`, conforming to the same decoder protocol as JSONL so they drop into `Stream`/`AsyncStream` unchanged
- Spec-faithful parsing per the WHATWG SSE spec: `CR` / `LF` / `CRLF` line splitting (incl. a lone `\r` split across chunks), `:`-comment lines ignored, one leading space stripped from values, multiple `data:` lines joined with `\n`, `id` persisted across events (NUL rejected), `retry` parsed only when ASCII-digits
- Always decoded as UTF-8: a single leading BOM is dropped only when present, and invalid byte sequences become U+FFFD instead of crashing the stream. JSONL stays strict (the JSONL spec forbids a BOM).

**Per package also:** sync + async tests, test-server JSONL/SSE routes, sample (`samples/sample_stream.py`), CHANGELOG entry + README section (azure-core also bumps version to 1.42.0).

> Automatic reconnection (honoring `retry` / re-sending `Last-Event-ID` on a dropped connection) is intentionally out of scope here — the decoders are pure parsers, and the reconnect loop belongs in a higher layer. Tracked as a follow-up.

### Validation (both packages)
- Streaming suite passes: **48 tests each** (sync + async), covering JSONL, SSE field/line-ending edge cases, BOM handling, and invalid-UTF-8 replacement

What client code to return a Stream() would look like:
image

### CODE REVIEW

File: sdk/core/azure-core/azure/core/streaming/_decoders.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    :rtype: Iterator[str]
+    :return: An iterator of lines.
+    """
+    decoder = codecs.getincrementaldecoder("utf-8")()
+
+    decoded = ""
+    for chunk in iter_bytes:
+        decoded += decoder.decode(chunk)
+        if decoded:
+            decoded_lines = decoded.splitlines()
```
Comment (kashifkhan) [inline sdk/core/azure-core/azure/core/streaming/_decoders.py]: I don't think we can use `splitlines` as is since it breaks on more than just what the JSONL spec specifies https://docs.python.org/3.4/library/stdtypes.html#str.splitlines. 

`splitlines` will treat `\u2028` and `\u2029` as split but the spec https://www.rfc-editor.org/info/rfc8259/ has it as 
```
 certain characters such as U+2028 LINE SEPARATOR and U+2029
   PARAGRAPH SEPARATOR are legal in JSON but not JavaScript
```

So line termination will have to be simply `\n` and `\r\n'

```python

'{"msg": "a\u2028b"}\n'.splitlines()
['{"msg": "a', 'b"}'  # one valid record -> two invalid fragments -> JSONDecodeError


'{"msg": "a\u2028b"}\n'.split("\n")
['{"msg": "a\u2028b"}', '']          # one record + trailing empty from the final \n
json.loads('{"msg": "a\u2028b"}')
{'msg': 'a\u2028b'}                   # U+2028 preserved inside the string value
```

File: sdk/core/azure-core/azure/core/streaming/_stream.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    ) -> None:
+        self._response = response
+        self._decoder = decoder
+        self._deserialization_callback = deserialization_callback
+        self._iterator = self._iter_results()
+
+    async def __anext__(self) -> ReturnType_co:
+        return await self._iterator.__anext__()
+
+    async def __aiter__(self) -> AsyncIterator[ReturnType_co]:  # pylint: disable=invalid-overridden-method
```
Comment (kashifkhan) [inline sdk/core/azure-core/azure/core/streaming/_stream.py]: I think the __iter__ funcs need to return `Self`

File: sdk/core/corehttp/corehttp/streaming/_stream.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        self._response.close()
+
+
+class AsyncStream(AsyncIterator[ReturnType_co]):
+    """AsyncStream class for asynchronously streaming JSONL.
+
+    :keyword response: The response object.
+    :paramtype response: ~corehttp.rest.AsyncHttpResponse
+    :keyword decoder: A decoder to use for the stream.
+    :paramtype decoder: ~corehttp.streaming.decoders.AsyncStreamDecoder
```
Comment (kashifkhan) [inline sdk/core/corehttp/corehttp/streaming/_stream.py]: should it be _decoders here?

File: sdk/core/corehttp/corehttp/streaming/_stream.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        self._iterator = self._iter_results()
+
+    def __next__(self) -> ReturnType_co:
+        return self._iterator.__next__()
+
+    def __iter__(self) -> Self:
+        return self
+
+    def _iter_results(self) -> Iterator[ReturnType_co]:
+        for event in self._decoder.iter_events(self._response.iter_bytes()):
```
Comment (l0lawrence) [inline sdk/core/corehttp/corehttp/streaming/_stream.py]: try/except here?

**Review by kashifkhan — APPROVED** [review]
